tools/llvm-pta-eval.cpp

```diff
@@ -0,0 +1,697 @@
+      score = "true";    // suspected
+    else if (aares == PartialAlias)
+      score = "inadequate";
+    else
+      score = "unknown";
+  } else if (fun.equals(EXPECTEDFAIL_NOALIAS)) {
+    r = (aares != NoAlias); ex = "EXPECTEDFAIL_NO";
+    if (aares == NoAlias)
+      score = "false";
+    else if (aares == MayAlias)
+      score = "true";
+    else if (aares == MustAlias)
+      score = "true";
+    else if (aares == PartialAlias)
+      score = "true";
+    else
+      score = "unknown";
+  } else {
+    return;
+  }
+
+  if (aares == NoAlias) s = "NO";
+  else if (aares == MayAlias) s = "MAY";
+  else if (aares == MustAlias) s = "MUST";
+  else s = "UNKNOWN";
+  if (r)
+    printf("  pta %s %s ex %s ", score, s, ex);
+  else
+    printf("  pta %s %s ex %s ", score, s, ex);
+
+  printf("\n");
+}
+
+static void
+evalPTA(LLVMPointerAnalysis *pta, PTType type)
+{
+
+  std::set<PSNode *> nodes;
+  pta->getNodes(nodes);
+
+  for (PSNode *node : nodes) {
+    evalPSNode(pta, node, type);
+  }
+}
+
+
+
+
+int main(int argc, char *argv[])
+{
+  llvm::Module *M;
+  llvm::LLVMContext context;
+  llvm::SMDiagnostic SMD;
+  bool todot = false;
+  const char *module = nullptr;
+  PTType type = FLOW_INSENSITIVE;
+  uint64_t field_senitivity = UNKNOWN_OFFSET;
+
+  // parse options
+  for (int i = 1; i < argc; ++i) {
+    // run given points-to analysis
+    if (strcmp(argv[i], "-pta") == 0) {
+      if (strcmp(argv[i+1], "fs") == 0)
+	type = FLOW_SENSITIVE;
+    } else if (strcmp(argv[i], "-pta-field-sensitive") == 0) {
+      field_senitivity = (uint64_t) atoll(argv[i + 1]);
+    } else if (strcmp(argv[i], "-dot") == 0) {
+      todot = true;
+    } else if (strcmp(argv[i], "-v") == 0) {
+      verbose = true;
+    } else {
+      module = argv[i];
+    }
+  }
+
+  if (!module) {
+    errs() << "Usage: % IR_module [output_file]\n";
+    return 1;
+  }
+
+#if ((LLVM_VERSION_MAJOR == 3) && (LLVM_VERSION_MINOR <= 5))
+  M = llvm::ParseIRFile(module, SMD, context);
+#else
+  auto _M = llvm::parseIRFile(module, SMD, context);
+  // _M is unique pointer, we need to get Module *
+  M = _M.get();
+#endif
+
+  if (!M) {
+    llvm::errs() << "Failed parsing '" << module << "' file:\n";
+    SMD.print(argv[0], errs());
+    return 1;
+  }
+
+  debug::TimeMeasure tm;
+
+  LLVMPointerAnalysis PTA(M, field_senitivity);
+  std::unique_ptr<PointerAnalysis> PA;
+
+  tm.start();
+
+  // use createAnalysis instead of the run() method so that we won't delete
+  // the analysis data (like memory objects) which may be needed
+  if (type == FLOW_INSENSITIVE) {
+    PA = std::unique_ptr<PointerAnalysis>(
+					  PTA.createPTA<analysis::pta::PointsToFlowInsensitive>()
+					  );
+  } else {
+    PA = std::unique_ptr<PointerAnalysis>(
+					  PTA.createPTA<analysis::pta::PointsToFlowSensitive>()
+					  );
+  }
+
+  // run the analysis
+  PA->run();
+
+  tm.stop();
+  tm.report("INFO: Points-to analysis [new] took");
+  dumpPointerSubgraph(&PTA, type, todot);
```

Comment:
is dumping the pointer subgraph still relevant here? If not, why don't we get rid of it (and all the associated code)